### PR TITLE
re-wire the response cache

### LINKED ISSUE #2297: CmdIDs and cross-range requests
I don't think we currently handle request idempotency correctly when operations cross ranges. The following could happen:

* client sets CmdID for a scan `[a,z)`.
* ranges `[a-q)`, `[q-z)` merge, but `DistSender` keeps a stale cache entry.
* the request gets split into two according to the stale descriptors and sent individually<sup>***</sup>
* responses are combined and a single response sent out, but there's a connection issue
* client retries, and by now DistSender has the stale entries removed so only one request is sent (with the client's `CmdID`).
* that request hits the response cache, but the entry cached is only the first of the two prior scans.

the problem is in <sup>***</sup>: For the second request we're sending, which `CmdID` do we use? Currently, it's the same for both requests - that of the client. But that's obviously wrong because the range handling both requests is the same in the above example, so the second request will get the first cached reply. If instead we create a new ID deterministically from the original one, we're still in trouble because it's not deterministic how many ranges the data is spread out over; in the above example the client's retry will hit the response cache, returning only half of its desired reply (`[a-q)`).

This looks like it's going to get complicated because even adding the key range to each response cache entry means there's a lot of subtlety involved because we'll have to recombine results from various cache entries for sub-keyranges.

One thing working in our favor is that (at least currently) we don't have many non-idempotent types of requests. Only `CPut`, `Inc` and `EndTransaction` actually need replay protection (at least in a Txn, which is always the case when `DistSender` thinks ranges are spanned). These are all point requests, so they're easier to handle. We still need to be careful about which `CmdID` to use for them: Due to the non-determinism of range descriptor lookup across retries we really have to use the original `CmdID` for all ranges, so the only discriminant is the key.

### PR DESCRIPTION
**Pull request by tbg**
removes `roachpb.ClientCmdID` and replay protection for non-transactional commands.
Instead, transactional commands are now cached on a per-ID response cache key. The cached value is the newly introduced _sequence number_ which is increased whenever a new request is sent as part of the transaction (in particular, in `client`, `DistSender`'s chunking sender and before `Store`-level retries).
Whenever a request encounters a cached sequence number greater than or equal to the transaction's, a `TransactionRetryError` is the result; the client will back of and come again.

In particular this fixes #2297 (since we don't return results, but simply an error) and #626 (since the response cache entries are now in 1:1 correspondence to transaction table entries, so #2062 is now the relevant issue covering this).

Commit-by-commit reviewing is suggested, though a casual look through the commit history probably makes sense to avoid comments invalidated by later changes.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3077)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r1, 7 of 7 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 5 of 5 files at r5, 6 of 6 files at r6, 1 of 1 files at r7, 8 of 8 files at r8, 12 of 12 files at r9, 21 of 21 files at r10.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

<sup>**[client/db_internal_test.go, line 37 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jOxqygGh5yqVWriDi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/client/db_internal_test.go#L37)):</sup>
include the numbers in the error


---

<sup>**[kv/txn_coord_sender.go, line 304 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jPEB5qb0Zp5BB4j4i)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/kv/txn_coord_sender.go#L304)):</sup>
has a sequence number?


---

<sup>**[roachpb/batch.go, line 262 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jPPG0hnLZ_pJbOcQ1)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/roachpb/batch.go#L262)):</sup>
Hm, alright then


---

<sup>**[storage/client_split_test.go, line 304 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jJ4kdmTMpfPYcWiEx)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e517fd63e1617af5716d948deaeb7885e3b8a4/storage/client_split_test.go#L304)):</sup>
include `err`?


---

<sup>**[storage/client_split_test.go, line 315 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jJ6xMOZqGhBnhNJK1)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e517fd63e1617af5716d948deaeb7885e3b8a4/storage/client_split_test.go#L315)):</sup>
include `err`?


---

<sup>**[storage/engine/rocksdb/db.cc, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jIQmRXQ0wJeCDOBNn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4a11cfce04a0ddecc4895ad87be3b1fd653239/storage/engine/rocksdb/db.cc#L186)):</sup>
This local can be removed now


---

<sup>**[storage/replica.go, line 1108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jJPraQQkpPhDuZl36)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e517fd63e1617af5716d948deaeb7885e3b8a4/storage/replica.go#L1108)):</sup>
dont see the value in this rename


---

<sup>**[storage/replica.go, line 143 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jPz0JjEJ273NQRc5V)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/storage/replica.go#L143)):</sup>
why not make this a struct instead of encoding all the stufff below?


---

<sup>**[storage/replica_test.go, line 1513 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jNefDcAb9iIvoFArn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1fc489c17ded3a0483b163aab9dfc67c12bcc9/storage/replica_test.go#L1513)):</sup>
can't you do equality or type assertion here?


---

<sup>**[storage/response_cache.go, line 174 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jKSElHu912B-pbsVp)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e517fd63e1617af5716d948deaeb7885e3b8a4/storage/response_cache.go#L174)):</sup>
The timestamp on the value is just so it matches the other argument? this function signature is pretty strange.


---

<sup>**[storage/response_cache.go, line 77 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jO16IEvah0Zju2ej0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1fc489c17ded3a0483b163aab9dfc67c12bcc9/storage/response_cache.go#L77)):</sup>
this comment is way stale


---

<sup>**[storage/response_cache.go, line 91 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jOVVKJMXP_tLnhTQv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1fc489c17ded3a0483b163aab9dfc67c12bcc9/storage/response_cache.go#L91)):</sup>
it should be mentioned somewhere that txn sequence are 1-indexed


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

File: storage/replica.go [hunk 1]
```diff
@@ -902,8 +901,15 @@ func (r *Replica) proposeRaftCommand(ctx context.Context, ba roachpb.BatchReques
 		OriginReplica: *replica,
 		Cmd:           ba,
 	}
-	cmdID := ba.GetOrCreateCmdID(r.store.Clock().PhysicalNow())
-	idKey := makeCmdIDKey(cmdID)
+
+	makeCmdIDKey := func() cmdIDKey {
+		buf := make([]byte, 0, 16)
+		buf = encoding.EncodeUint64(buf, uint64(r.store.Clock().PhysicalNow()))
```
Comment (bdarnell) [inline storage/replica.go]: Now that `cmdIDKeys` are not used as rocksdb keys, I don't think we need to have a time prefix. They can just be random.

**Comment by bdarnell** [comment]
LGTM

**Comment by petermattis** [comment]
So much code deletion!


---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[storage/replica.go, line 907 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jHVSe4BSpQ0h_uth2-r10-907)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/storage/replica.go#L907)):</sup>
Why even random? Seems like you could have a sequence number inside `Replica`. That seems preferable to a random number for debugging purposes.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[kv/txn_coord_sender.go, line 304 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jPEB5qb0Zp5BB4j4i)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/kv/txn_coord_sender.go#L304)):</sup>
nope, tracing uses only the transaction ID.


---

<sup>**[storage/replica.go, line 1108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jJPraQQkpPhDuZl36)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e517fd63e1617af5716d948deaeb7885e3b8a4/storage/replica.go#L1108)):</sup>
`if err := func(err)` looks mildly confusing.


---

<sup>**[storage/replica.go, line 143 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jPz0JjEJ273NQRc5V)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/storage/replica.go#L143)):</sup>
MultiRaft uses a string internally.


---

<sup>**[storage/replica_test.go, line 1513 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jNefDcAb9iIvoFArn)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1fc489c17ded3a0483b163aab9dfc67c12bcc9/storage/replica_test.go#L1513)):</sup>
no, the replica corruption error is package-internal, and this is in storage_test.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by tschottdorf** [comment]
added some commits addressing comments, PTAL.

**Comment by tamird** [comment]
Reviewed 12 of 12 files at r11, 3 of 3 files at r12, 3 of 3 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[roachpb/api.go, line 54 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2n4Of2ggzNYrO2NLHh)** ([raw file](https://github.com/cockroachdb/cockroach/blob/769d2cc4aa41edc4958bef994b02bc4c2977d500/roachpb/api.go#L54)):</sup>
instead of a map, how about `[...]string{`? You'll need to put the sentinel in the array.


---

<sup>**[roachpb/api.go, line 506 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2n4Z9GJMq_PXWHW8s-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/769d2cc4aa41edc4958bef994b02bc4c2977d500/roachpb/api.go#L506)):</sup>
use a buffer?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[storage/replica.go, line 907 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jHVSe4BSpQ0h_uth2-r10-907)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/storage/replica.go#L907)):</sup>
Export `multiraft.commandIDLen` instead of hard-coding 16 here. We can also change that constant to 8 now that our command IDs are smaller.

I'm not sure if it's safe to make this a per-`Replica` sequence. In multiraft the ids need to be unique per `multiraft.group`, which mostly correspond to Replicas but have slightly different lifetimes. I'd rather keep it random.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[roachpb/api.go, line 54 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2n4Of2ggzNYrO2NLHh)** ([raw file](https://github.com/cockroachdb/cockroach/blob/769d2cc4aa41edc4958bef994b02bc4c2977d500/roachpb/api.go#L54)):</sup>
that would work for `flagToStr`, but it would make for a rather awkward definition (the flags are 2^i).


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by tschottdorf** [comment]
PTAL baby one more time

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r15.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[storage/replica.go, line 907 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077#-K2jHVSe4BSpQ0h_uth2-r10-907)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e2e3f6b0b1ef7031a4453dbef76a1c2a635f9784/storage/replica.go#L907)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3077)

**Pull request merged**